### ISSUE TITLE
[BUG] merge with sort=True sometimes doesn't

### ISSUE THREAD

**Issue by lmeyerov** [rapidsai/cudf#4374]
**Describe the bug**

Left `merge` with `sort=True` doesn't sort, sometimes.

**Steps/Code to reproduce bug**

Run this a few times:

```
_________________________________________________________________________________________________ test_merge_sort _________________________________________________________________________________________________

    def test_merge_sort():
        left = cudf.DataFrame({
            'id': [0, 1]
        })
        right = cudf.DataFrame({
            'id': [1],
            'v': ['a']
        })
        merged = left.merge(right, on='id', how='left', sort=True).reset_index(drop=True)
        manual = left.merge(right, on='id', how='left').sort_values(by='id').reset_index(drop=True)
        logger.trace('--------------')
        logger.trace(('merged'))
        logger.trace(merged.to_pandas())
        logger.trace('--------------')
        logger.trace(('manual'))
        logger.trace((manual.to_pandas()))
        logger.trace('--------------')
>       assert merged.to_pandas().to_dict() == manual.to_pandas().to_dict()
E       AssertionError: assert {'id': {0: 1,...'a', 1: None}} == {'id': {0: 0,...None, 1: 'a'}}
E         Differing items:
E         {'v': {0: 'a', 1: None}} != {'v': {0: None, 1: 'a'}}
E         {'id': {0: 1, 1: 0}} != {'id': {0: 0, 1: 1}}
E         Full diff:
E         - {'id': {0: 1, 1: 0}, 'v': {0: 'a', 1: None}}
E         ?            ^     ^           --------
E         + {'id': {0: 0, 1: 1}, 'v': {0: None, 1: 'a'}}...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show

test/server/tasks/util/test_shape_graph_clean.py:67: AssertionError
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
TRACE:--------------
TRACE:merged
TRACE:   id     v
0   1     a
1   0  None
TRACE:--------------
TRACE:manual
TRACE:   id     v
0   0  None
1   1     a
TRACE:--------------
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
TRACE    tasks.util.shaping:test_shape_graph_clean.py:60 --------------
TRACE    tasks.util.shaping:test_shape_graph_clean.py:61 merged
TRACE    tasks.util.shaping:test_shape_graph_clean.py:62    id     v
0   1     a
1   0  None
TRACE    tasks.util.shaping:test_shape_graph_clean.py:63 --------------
TRACE    tasks.util.shaping:test_shape_graph_clean.py:64 manual
TRACE    tasks.util.shaping:test_shape_graph_clean.py:65    id     v
0   0  None
1   1     a
TRACE    tasks.util.shaping:test_shape_graph_clean.py:66 --------------
```

**Expected behavior**

In the above, the expected output order is to be sorted on left-key `id`, yielding id order of `0`, `1`.

**Environment overview (please complete the following information)**

RAPIDS 0.12 in docker, but an ~old GPU.

**Environment details**

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 435.21       Driver Version: 435.21       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce MX150       Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   39C    P8    N/A /  N/A |      0MiB /  2002MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
```

**Comment by kkraus14** [comment]
@brandon-b-miller I believe you're fixing this in #4466

**Comment by lmeyerov** [comment]
Thanks, I didn't even realize Pandas did that -- see comment on why it negates the use of `sort=True` in practice: https://github.com/rapidsai/cudf/pull/4466#issuecomment-598870951

**Comment by kkraus14** [comment]
Fixed in #4466
